client/src/fuzzy-search.ts

```diff
@@ -1,124 +1,57 @@
-    // is to use `fuzzysearch()`. It's significantly faster than anything
-    // from `fuzzyjs`.
-    // So first filter out the "junk" by only considering those that match
-    // at all.
-    const candidates = this.haystack
-      .filter((doc) => {
-        // You could pre-compute a list of just the URLs all in lowercase,
-        // but through experimentation, it's clear that this transformation
-        // is sufficiently fast. And it's more memory efficient because
-        // you don't need to store that second array.
-        return fuzzysearch(needleLower, doc.url.toLowerCase());
-      })
-      // We have to cap the number of candidates because `fuzzyjs` is slow.
-      // Remember that the haystack is naturally sorted by popularity
-      // which gives the most populary an upper hand here.
-      // The number is quite arbitrary. Through careful experimentation
-      // it's found that it takes about 25ms to fully sort and filter
-      // for 1,000 candidates.
-      // Also, this compromise is justified in that if there that many
-      // candidates to begin with the needle is bound to match too much
-      // so the benefit of sorting by popularity is fair.
-      .slice(0, 1000);
-
-    // Now that we have a subset of only those that match, we can proceed
-    // and use `fuzzyjs` to sort by its sorting function and extract
-    // each match's ranges so we can finally result a list of sub-strings.
-    const res: CombinedMatchResult[] = candidates
-      .map((doc) => {
-        const matched = match(needle, doc.url, {
-          withScore: true,
-          withRanges: true,
-        });
-        return Object.assign(matched, doc);
-      })
-      .filter((m) => m.match && m.score && m.score > 0)
-      .sort((a, b) => {
-        // TypeScript doesn't know what we used `withScores: true`
-        // so have to do this if statement every time.
-        if (b.score && a.score) {
-          return b.score - a.score;
+  search(needle: string, { limit = 10 }): FzfResultItem<Doc>[] {
+    // Use `let` because we might come up with a new list (aka. shortlist)
+    // that makes the haystack search much simpler to send to `Fzf()`.
+    let docs = this.docs;
+    // The list of docs is possible over 10,000 entries (in 2021). If the
+    // search input is tiny, don't bother with the overhead of Fzf().
+    // Because we don't even need it when the test is so easy in that
+    // it just needs to contain a single character.
+    if (needle.length <= 3) {
+      const needleLowerCase = needle.toLowerCase();
+      // The reason this works and makes sense is because the `this.docs` is
+      // already sorted by popularity.
+      // So if someone searches for something short like `x` we just take
+      // the top 'limit' docs that have an `x` in the `.url`. This is
+      // faster than going through
+      const shortlistDocs: Doc[] = [];
+      for (const doc of this.docs) {
```

Comment:
This makes sense to me! Since we know `this.docs` will be an array of `url` and `title` objects and that its sorted by popularity, there isn't a reason to do another clean sweep of the 10k+ docs when we can simply define an upper bound (limit) for tiny search inputs and push them into the `shortlistDocs` array of `Doc` objects. LGTM!